src/fd.cpp

```diff
@@ -0,0 +1,264 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Increase max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+
+#endif
+
+pthread_attr_t pt_attr;
+int pt_attr_created = 0;
```

Comment:
I don't think this is worth the global variable, I think the attribute objects are super cheap to create/destroy. And you can destroy as soon as the thread is created.

Reply:
Actually, is there a reason you're not using tinycthread to launch new threads? (`tct_thrd_create()`)

Reply:
Precisely this ability to launch a thread with detached state set! Probably being too paranoid about race conditions (thread finished before `tct_thread_detach()` is called). Also there's no cancellation API.

It's a really thin wrapper - I think it's pretty much equivalent me using tct and writing the missing methods, or calling the underlying functions as I'm doing now.

Reply:
> I don't think this is worth the global variable, I think the attribute objects are super cheap to create/destroy. And you can destroy as soon as the thread is created.

I've reverted this in [e4486fc](https://github.com/r-lib/later/pull/190/commits/e4486fc70a85593d60488c4db0ab92d6c21577ce). Also simplified a bit with a typedef for the platform-specific thread types.

Reply:
(Sorry, I seem to have missed a number of your comments that you wrote yesterday, which is why I've made some comments today that were already answered by your previous comments.)

Reply:
My turn to be paranoid :) I hadn't gotten to the TerminateThread/pthread_cancel part, I always had it drilled into me that preemptive termination like that is unsafe in any situation (except sometimes if the entire process is going down, and even then pretty sketchy). I would either have a cancellation flag and call select in a loop with a shorter timeout, or else, use a pipe for signaling purposes.

Can the worker threads call `tct_thread_detach` on themselves upon start? I don't think there's any race condition then.

Reply:
You're right to be paranoid in your case! I had planned to upgrade to using a pipe when I introduced the re-usable thread (otherwise would seem to be too costly), but your idea of a cancellation flag is even better!
I've implemented this using a `std::atomic<bool>` in [bf48672](https://github.com/shikokuchuo/later/commit/bf48672b9cfa52470276a37a47271b55144e0c48).

When I do the re-usable thread, I'll break down the poll to check in [400ms] intervals (it's how I do interruptible timed waits for condition variables in nanonext), but for now it works well enough without the added complexity. If flagged, the callback is not executed, but importantly the callback is still released.

The behaviour of the cancellation function now also matches that of `later()` i.e. returns TRUE if cancelled, and FALSE subsequently, or if the callback has already run.

Reply:
> Can the worker threads call `tct_thread_detach` on themselves upon start? I don't think there's any race condition then.

This should work, but there's also a race condition as to whether the detach happens before the launching function terminates - and I'm not certain what happens in that case.

I've re-read the documentation for POSIX threads, and it seems safe to just call detach immediately after thread creation. If the thread has already terminated then the function should just return an error (which we ignore).

Reply:
> Actually, is there a reason you're not using tinycthread to launch new threads? (`tct_thrd_create()`)

Just to address all of the comments here: I've completely transitioned to tinycthread now.